src/org/labkey/targetedms/MsDataSourceUtil.java

```diff
@@ -0,0 +1,1243 @@
+import org.labkey.api.data.dialect.SqlDialect;
+import org.labkey.api.exp.ExperimentException;
+import org.labkey.api.exp.Lsid;
+import org.labkey.api.exp.api.DataType;
+import org.labkey.api.exp.api.ExpData;
+import org.labkey.api.exp.api.ExperimentService;
+import org.labkey.api.files.FileContentService;
+import org.labkey.api.module.ModuleLoader;
+import org.labkey.api.query.FieldKey;
+import org.labkey.api.security.User;
+import org.labkey.api.targetedms.ISampleFile;
+import org.labkey.api.targetedms.TargetedMSService;
+import org.labkey.api.util.FileUtil;
+import org.labkey.api.util.JunitUtil;
+import org.labkey.api.util.TestContext;
+import org.labkey.targetedms.parser.Instrument;
+import org.labkey.targetedms.parser.PsiInstruments;
+import org.labkey.targetedms.parser.SampleFile;
+import org.labkey.targetedms.query.InstrumentManager;
+
+import java.io.IOException;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import static org.hamcrest.CoreMatchers.hasItem;
+
+public class MsDataSourceUtil
+{
+    private static final MsDataSourceUtil instance = new MsDataSourceUtil();
+    private static final Logger LOG = Logger.getLogger(MsDataSourceUtil.class);
+
+    private MsDataSourceUtil() {}
+
+    public static MsDataSourceUtil getInstance()
+    {
+        return instance;
+    }
+
+    public RawDataInfo getDownloadInfo(@NotNull SampleFile sampleFile, @NotNull Container container)
+    {
+        ExperimentService expSvc = ExperimentService.get();
+        if(expSvc == null)
+        {
+            return null;
+        }
+
+        // We will look for raw data files only in @files/RawFiles
+        Path rawFilesDir = getRawFilesDir(container);
+        if(rawFilesDir == null)
+        {
+            return null;
+        }
+
+        ExpData expData = getDataForSampleFile(sampleFile, container, rawFilesDir, expSvc, false);
+
+        Long size;
+        if(expData != null)
+        {
+            Path dataPath = expData.getFilePath();
+            if(dataPath != null && Files.exists(dataPath))
+            {
+                try
+                {
+                    if(!Files.isDirectory(dataPath))
+                    {
+                        size = Files.size(dataPath);
+                        return new RawDataInfo(expData, size, true);
+                    }
+                    else
+                    {
+                        size = FileUtil.hasCloudScheme(dataPath) ? null : FileUtils.sizeOfDirectory(dataPath.toFile());
+                        return new RawDataInfo(expData, size, false);
+                    }
+                }
+                catch (IOException e)
+                {
+                    LOG.debug("Error getting size of " + dataPath, e);
```

Comment:
IOExceptions indicate there's something going wrong with the file system. debug() level messages will be effectively ignored via the logging configuration by default. I recommend throwing an UnexpectedException instead, wrapping the original IOException.

If there are specific cases where you want the error to be non-fatal, please log at least at the warn level so that it's more likely an admin will see it and correct the problem

Reply:
Done.